graph/api/api_perf_test.go

```diff
@@ -0,0 +1,1178 @@
+		v12 := model.Vector{}
+
+		q13 := fmt.Sprintf(`round(sum(rate(istio_response_messages_total{reporter="destination",destination_workload_namespace="bookinfo%d"} [600s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision) > 0,0.001)`, i)
+		v13 := model.Vector{}
+
+		q14 := fmt.Sprintf(`round(sum(rate(istio_response_messages_total{reporter="source",source_workload_namespace="bookinfo%d"} [600s])) by (source_cluster,source_workload_namespace,source_workload,source_canonical_service,source_canonical_revision,destination_cluster,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_canonical_service,destination_canonical_revision) > 0,0.001)`, i)
+		q14m0 := model.Metric{
+			"source_workload_namespace":      model.LabelValue(fmt.Sprintf("bookinfo%d", i)),
+			"source_workload":                "reviews-v3",
+			"source_canonical_service":       "reviews",
+			"source_canonical_revision":      "v3",
+			"source_cluster":                 "east",
+			"destination_cluster":            "east",
+			"destination_service_namespace":  "bankapp",
+			"destination_service":            "deposit:9080",
+			"destination_service_name":       "deposit",
+			"destination_workload_namespace": "bankapp",
+			"destination_workload":           "deposit-v1",
+			"destination_canonical_service":  "deposit",
+			"destination_canonical_revision": "v1",
+		}
+		v14 := model.Vector{
+			&model.Sample{
+				Metric: q14m0,
+				Value:  100,
+			},
+		}
+
+		mockQuery(api, q6, &v6)
+		mockQuery(api, q7, &v7)
+		mockQuery(api, q8, &v8)
+		mockQuery(api, q9, &v9)
+		mockQuery(api, q10, &v10)
+		mockQuery(api, q11, &v11)
+		mockQuery(api, q12, &v12)
+		mockQuery(api, q13, &v13)
+		mockQuery(api, q14, &v14)
+	}
+
+	return client, api, nil, biz
+}
+
+func BenchmarkVersionedAppGraph(b *testing.B) {
+	parseFlags(b)
+
+	client, _, err, biz := mockNamespaceRatesGraphPerf(b, numberOfNamespacesFlag)
+	if err != nil {
+		b.Fatal(err)
+	}
+
+	namespaces := ""
+	for i := 1; i <= numberOfNamespacesFlag; i++ {
+		if i > 1 {
+			namespaces += ","
+		}
+		namespaces += fmt.Sprintf("bookinfo%d", i)
+	}
+
+	mr := mux.NewRouter()
+	mr.HandleFunc("/api/namespaces/graph", func(w http.ResponseWriter, r *http.Request) {
+		options := graph.NewOptions(r, &biz.Namespace)
+		options.Rates.Ambient = graph.AmbientTrafficNone
+		code, config := graphNamespacesIstio(r.Context(), biz, client, options)
+		respond(w, code, config)
+	})
```

Comment:
Can you add appenders to the query? Since we're mocking out the prom queries here the two areas this benchmark will help in optimizing are appenders and the graph gen converting prom metrics --> in memory graph structure. Whatever options the frontend passes in the URL would be good to pass here. iirc all appenders are run by default.